### PR TITLE
Automatic Percolation-Based Length Scale for Clipper System

### PR DESCRIPTION
**Pull request by mstormbull**
## Summary

Adds automatic calculation of the `length_scale` parameter in `ClipperConfig` based on percolation theory for uniformly sampled 2D space. The length scale is now automatically computed from grid dimensions and building density, eliminating manual configuration while providing a theoretically grounded default.

## Motivation

The clipper system's spatial spread behavior is controlled by `length_scale`, which determines how clipping propagates across the grid via exponential decay (`weight = exp(-distance / length_scale)`). Previously, this required manual tuning. 

This PR implements automatic calculation using the **critical percolation threshold (4.51)** for 2D continuum percolation, providing a principled default that adapts to:
- Grid size (larger grids → larger length scales)
- Building density (more buildings → smaller length scales)

Note that this calculation holds in principle only for uniform distributions, but the configurable fudge factor can be applied to safely use this to scale to different types of resource clustering. 

### Core Implementation

**Modified: `metta/packages/mettagrid/python/src/mettagrid/config/mettagrid_config.py`**

1. Added new fields to `ClipperConfig`:
   - `auto_length_scale: bool = True` - Enable/disable automatic calculation
   - `length_scale_factor: float = 1.0` - Tunable fudge factor for the theoretical value

2. Added model validator to `GameConfig`:
   - `_auto_configure_clipper_length_scale()` - Automatically calculates length scale on config initialization
   - Formula: `length_scale = (grid_size / √num_buildings) × √(4.51 / 4π) × fudge_factor`
   - Gracefully handles edge cases (no clipper, no buildings, missing dimensions)

### Testing

**Created: `metta/packages/mettagrid/tests/test_clipper_percolation.py`**
- 8 comprehensive tests, all passing
- Tests automatic calculation, fudge factor tuning, manual override, edge cases
- Verifies correct scaling with grid size and building density

### Documentation

**Created: `metta/packages/mettagrid/CLIPPER_PERCOLATION_QUICK_START.md`**
- Quick reference guide with usage examples
- Parameter descriptions and example values

**Updated: `/PERCOLATION_LENGTH_SCALE_IMPLEMENTATION.md`**
- Full implementation details and theory background

### Default (Automatic)

```python
game_config = GameConfig(
    map_builder=RandomMapBuilder.Config(width=50, height=50),
    objects={
        f"assembler_{i}": AssemblerConfig(name=f"assembler_{i}", type_id=i+1)
        for i in range(25)
    },
    clipper=ClipperConfig(
        recipe=RecipeConfig(input_resources={"ore_red": 5}),
        clip_rate=0.1,
        # length_scale automatically calculated → ~5.991
    ),
)
```

### With Tuning

```python
clipper=ClipperConfig(
    recipe=RecipeConfig(input_resources={"ore_red": 5}),
    clip_rate=0.1,
    length_scale_factor=1.5,  # Scale the theoretical value up or down (will be equal for all density)
)
```

### Manual Override

```python
clipper=ClipperConfig(
    recipe=RecipeConfig(input_resources={"ore_red": 5}),
    clip_rate=0.1,
    auto_length_scale=False,  # Disable automatic
    length_scale=3.14,  # Use manual value
)
```

## Example Values

**50×50 grid, 25 buildings:**
- `fudge_factor=0.5` → `length_scale=2.995` (localized)
- `fudge_factor=1.0` → `length_scale=5.991` (theoretical)
- `fudge_factor=1.5` → `length_scale=8.986` (diffuse)
- `fudge_factor=2.0` → `length_scale=11.982` (very diffuse)

**50×50 grid, varying density (fudge=1.0):**
- 10 buildings → `length_scale=9.472`
- 25 buildings → `length_scale=5.991`
- 50 buildings → `length_scale=4.236`
- 100 buildings → `length_scale=2.995`

## Testing

```bash
cd metta/packages/mettagrid
uv run pytest tests/test_clipper_percolation.py -v
```

**Results:** ✅ 8/8 tests passing

## Backward Compatibility

✅ **Fully backward compatible**
- Existing configs without `clipper` continue to work
- Manual `length_scale` values are preserved when `auto_length_scale=False`
- Default behavior is automatic calculation, but can be disabled

## Benefits

1. **Zero configuration required** - Works automatically out of the box
2. **Theoretically grounded** - Based on percolation theory (threshold ≈ 4.51)
3. **Adaptive** - Automatically scales with grid size and building density
4. **Tunable** - `length_scale_factor` allows experimentation
5. **Flexible** - Can override with manual values when needed

## Files Changed

```
metta/packages/mettagrid/python/src/mettagrid/config/mettagrid_config.py
metta/packages/mettagrid/tests/test_clipper_percolation.py (new)
metta/packages/mettagrid/CLIPPER_PERCOLATION_QUICK_START.md (new)
PERCOLATION_LENGTH_SCALE_IMPLEMENTATION.md (new)
```

## Checklist

- ✅ Implementation complete
- ✅ All tests passing (8/8)
- ✅ Code formatted with ruff
- ✅ Documentation written
- ✅ Backward compatible
- ✅ No breaking changes

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1211527187906777)

### CODE REVIEW

File: packages/mettagrid/python/src/mettagrid/config/mettagrid_config.py [hunk 1]
```diff
@@ -275,6 +299,40 @@ def _coerce_map_builder(cls, value: Any) -> Any:
...
+
+        if width is None or height is None:
+            # Can't auto-calculate, leave length_scale as-is
+            return self
+
+        # Calculate grid size (use max dimension)
+        grid_size = max(width, height)
+
+        # Count assemblers in objects
+        num_buildings = sum(1 for obj in self.objects.values() if isinstance(obj, AssemblerConfig))
```
Comment (sasmith) [inline packages/mettagrid/python/src/mettagrid/config/mettagrid_config.py:323]: This is going to give you the wrong number. We should expect to have fewer object configs than we have objects on the map.

I think whatever calculation we want to do needs to be post map generation.

**Review by sasmith — CHANGES_REQUESTED** [review]
We discussed moving this to be at the cpp layer, so handing back.

**Comment by mstormbull** [comment]
@sasmith thanks for your patience. Ready for review.
